File: longestConsecutive.cpp
//hashset
//time complexity -> O(n) due to using for loop + while loop
//space compelxity -> O(n) due to using hashset
#include <unordered_set>
#include <vector>

using namespace std;
class Solution{
public:
    int longestConsecutive(vector<int>& nums){
        unordered_set<int> hashSet(nums.begin(), nums.end());
        int longest = 0;
        for(int num : hashSet){
            if (hashSet.find(num-1) == hashSet.end()){
                int length = 1;
                while(hashSet.find(num+length) != hashSet.end()){
                    length++;
                }
                longest = max(length,longest);
            }
        }
        return longest;
    }
};
///////////////////////////////////////////////////////
///////////////////////////////////////////////////////

File: longestConsecutive.py
#hashset
#time complexity -> O(n) because of for loop + while
#space complexity -> O(n) in memory because uses hashset as a data structure
from typing import List
class Solution:
    def longestConsecutive(self, nums: List[int]):
        hashset = set(nums)
        
        longest = 0
        for num in hashset:
            if num-1 in hashset:
                length = 1
                while (num + length) in hashset:
                    length += 1
                longest = max(longest, length)

        return longest

#######################################################
#######################################################

